cleo_plugins/Audio/CSoundSystem.cpp

```diff
@@ -0,0 +1,459 @@
+    {
+        unsigned flags = BASS_SAMPLE_SOFTWARE;
+        if (CSoundSystem::useFloatAudio) flags |= BASS_SAMPLE_FLOAT;
+
+        if (!(streamInternal = BASS_StreamCreateFile(FALSE, src, 0, 0, flags)) &&
+            !(streamInternal = BASS_StreamCreateURL(src, 0, flags, 0, nullptr)))
+        {
+            LOG_WARNING(0, "Loading audiostream %s failed. Error code: %d", src, BASS_ErrorGetCode());
+            return;
+        }
+        
+        BASS_ChannelGetAttribute(streamInternal, BASS_ATTRIB_FREQ, &rate);
+        OK = true;
+    }
+
+    CAudioStream::~CAudioStream()
+    {
+        if (streamInternal) BASS_StreamFree(streamInternal);
+    }
+
+    C3DAudioStream::C3DAudioStream(const char *src) : CAudioStream()
+    {
+        unsigned flags = BASS_SAMPLE_3D | BASS_SAMPLE_MONO | BASS_SAMPLE_SOFTWARE;
+        if (CSoundSystem::useFloatAudio) flags |= BASS_SAMPLE_FLOAT;
+
+        if (!(streamInternal = BASS_StreamCreateFile(FALSE, src, 0, 0, flags)) &&
+            !(streamInternal = BASS_StreamCreateURL(src, 0, flags, nullptr, nullptr)))
+        {
+            LOG_WARNING(0, "Loading 3d-audiostream %s failed. Error code: %d", src, BASS_ErrorGetCode());
+            return;
+        }
+
+        BASS_ChannelGetAttribute(streamInternal, BASS_ATTRIB_FREQ, &rate);
+        BASS_ChannelSet3DAttributes(streamInternal, BASS_3DMODE_NORMAL, 3.0f, 1E+12f, -1, -1, -1.0f);
+        OK = true;
+    }
+
+    C3DAudioStream::~C3DAudioStream()
+    {
+        if (streamInternal) BASS_StreamFree(streamInternal);
+    }
+
+    void CAudioStream::Play()
+    {
+        Process(true); // update settings
+        BASS_ChannelPlay(streamInternal, TRUE);
+        state = playing;
+    }
+
+    void CAudioStream::Pause(bool change_state)
+    {
+        BASS_ChannelPause(streamInternal);
+        if (change_state) state = paused;
+    }
+
+    void CAudioStream::Stop()
+    {
+        BASS_ChannelPause(streamInternal);
+        BASS_ChannelSetPosition(streamInternal, 0, BASS_POS_BYTE);
+        state = paused;
+    }
+
+    void CAudioStream::Resume()
+    {
+        BASS_ChannelPlay(streamInternal, FALSE);
+        state = playing;
+    }
+
+    float CAudioStream::GetLength() const
+    {
+        return (float)BASS_ChannelBytes2Seconds(streamInternal, BASS_ChannelGetLength(streamInternal, BASS_POS_BYTE));
+    }
+
+    DWORD CAudioStream::GetState() const
+    {
+        if (state == stopped) return -1; // dont do this in case we changed state by pausing
+
+        switch (BASS_ChannelIsActive(streamInternal))
+        {
+            case BASS_ACTIVE_STOPPED:
+            default:
+                return -1;
+            case BASS_ACTIVE_PLAYING:
+            case BASS_ACTIVE_STALLED:
+                return 1;
+            case BASS_ACTIVE_PAUSED:
+                return 2;
+        };
+    }
+
+    void CAudioStream::SetVolume(float val, float transitionTime)
+    {
+        if (val < 0.0f) // just update
```

Comment:
lets make a separate method `UpdateVolume()` instead of calling SetVolume(-1.0f). it makes the code easier to understand

Reply:
Done.